---
title: "An introduction to ReadAb"
author: "Clare Gillis"
date: "`r format(Sys.time(), '%d %b %Y')`"
output: 
  rmarkdown::html_vignette:
    toc: true
    number_sections: false
vignette: >
  %\VignetteIndexEntry{An introduction to ReadAb}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
library(knitr)
opts_chunk$set(fig.align = "center", 
               out.width = "90%",
               fig.width = 6, fig.height = 5.5,
               dev.args=list(pointsize=10),
               par = TRUE, # needed for setting hook 
               collapse = TRUE, # collapse input & ouput code in chunks
               warning = FALSE)
knit_hooks$set(par = function(before, options, envir)
  { if(before && options$fig.show != "none") 
       par(family = "sans", mar=c(4.1,4.1,1.1,1.1), mgp=c(3,1,0), tcl=-0.5)
})
set.seed(1) # for exact reproducibility
library(ReadAb)
```

## Introduction

`ReadAb` is an R package developed for reading antibody structures, identifying
CDRs (complimentarity determining regions, AKA binding loops), and visualizing and performing 
analysis on these structures.

To download **ReadAb**, use the following commands:

``` r
require("devtools")
devtools::install_github("claregillis/ReadAb", build_vignettes = TRUE)
library("ReadAb")
```

To list all sample datasets available in the package:
``` r
data(package = "ReadAb")
```

<br>

## Functionality

### 1. Read an Antibody Structure

To begin, you need to load antibody structures using the `ReadAntibody` function from the `ReadAb` package. This function requires the path to the PDB file, the numbering scheme used for loop locations, and the chain identifiers for heavy, light, and antigen chains.

Note: you can find the identifiers for heavy, light, and antigen chains at the 
top of the PDB.

#### Example: Reading the Antibody Structure

```{r}
# Load the antibody structure
pdbPath <- system.file("extdata", "7x94_imgt.pdb", package = 'ReadAb')
ab7x94 <- ReadAb::ReadAntibody(pdb = pdbPath,
                          numbering = "IMGT",
                          heavy = "H",
                          light = "L",
                          antigen = "A")
```

This command loads the 7x94_imgt.pdb structure and stores it in the ab7x94 object. You can then access specific loops using the following code:

```{r}
# Access the H1 loop
H1 <- ab7x94$loops$H1
```

### 2. Assess Loop Similarity

After loading the antibody structures, you can compare the similarity of specific loops using the `AssessLoopSimilarity` and `AssessOverallLoopSimilarity` functions.

#### Comparing a Single Loop

To compare the H1 loops between two antibodies (ab7x94 and ab7x96):

```{r}
# Load a second antibody structure
ab7x96 <- ReadAb::ReadAntibody(pdb = system.file("extdata", "7x96_imgt.pdb", package = "ReadAb"),
                          numbering = "IMGT",
                          heavy = "H",
                          light = "L",
                          antigen = "A")

# Compare H1 loops
H1Similarity <- ReadAb::AssessLoopSimilarity(list(ab7x94, ab7x96), 'H1')
```

#### Comparing All Loops with Weights

To compare all loops of the two antibodies and assign different weights to each loop, use `AssessOverallLoopSimilarity.` For example, to give the H3 loop a higher weight:

```{r}
# Compare all loops with weighted similarity
OverallSimilarity <- ReadAb::AssessOverallLoopSimilarity(
  list(ab7x94, ab7x96),
  wH1 = 0.1, wH2 = 0.1, wH3 = 0.5,
  wL1 = 0.1, wL2 = 0.1, wL3 = 0.1
)
```

### 3. Visualize the Antibody Structure

The `VisualizeAntibody` function allows you to view the 3D structure of an antibody and color different components.

#### Displaying All Atoms

To view all atoms of the antibody ab7x94:

```{r}
# Visualize all atoms of the antibody
ReadAb::VisualizeAntibody(antibody = ab7x94, mode = "all_atoms")
```

#### Displaying Backbone Atoms and Changing Color

You can focus on the backbone atoms and color specific components (e.g., antigen) for better visualization:

```{r}
# Color antigen and display backbone atoms
cyanAgAb7x94 <- ReadAb::SetComponentColor(ab7x94, 'antigen', '#42f5dd')
ReadAb::VisualizeAntibody(cyanAgAb7x94, mode = "backbone")
```

### 4. Visualize Loop Similarity with Heatmaps

The similarity results from the `AssessLoopSimilarity` and `AssessOverallLoopSimilarity` functions can be visualized as heatmaps using the `DisplaySimilarityPlot` function.

#### Displaying a Heatmap for H3 Loop Similarity

To visualize the similarity of H3 loops between the two antibodies:

```{r}
# Visualize the similarity of H3 loops
H3Similarity <- ReadAb::AssessLoopSimilarity(list("7x94" = ab7x94, "7x96" = ab7x96), 'H3')
ReadAb::DisplaySimilarityPlot(H3Similarity, 'H3')
```

### 5. Comparing Multiple Antibodies

You can extend this process to compare four antibodies. Load the structures of all four antibodies and assess the similarity of each loop.

#### Example: Comparing H1 Loops of Four Antibodies

```{r}
# Load four antibodies
ab1 <- ReadAb::ReadAntibody(
  pdb = system.file("extdata", "7x94_imgt.pdb", package = "ReadAb"),
  numbering = "IMGT",
  heavy = "H",
  light = "L",
  antigen = "A"
)
ab2 <- ReadAb::ReadAntibody(
  pdb = system.file("extdata", "7x96_imgt.pdb", package = "ReadAb"),
  numbering = "IMGT",
  heavy = "H",
  light = "L",
  antigen = "A"
)
ab3 <- ReadAb::ReadAntibody(
  pdb = system.file("extdata", "7uja_chothia.pdb", package = "ReadAb"),
  numbering = "Chothia",
  heavy = c("B", "E", "G", "I", "L", "N"),
  light = c("D", "F", "H", "J", "M", "O"),
  antigen = c("A", "K", "C")
)
ab4 <- ReadAb::ReadAntibody(
  pdb = system.file("extdata", "7ru4_imgt.pdb", package = "ReadAb"),
  numbering = "IMGT",
  heavy = "H",
  light = "L",
  antigen = "A"
)

# Compare H1 loops
H1Similarity <- ReadAb::AssessLoopSimilarity(list("Antibody1" = ab1, 
                                                  "Antibody2" = ab2, 
                                                  "Antibody3" = ab3, 
                                                  "Antibody4" = ab4), 
                                             'H1')

# Display the similarity plot
ReadAb::DisplaySimilarityPlot(H1Similarity, loop = 'H1')
```

#### Example: Overall Similarity of All Loops

You can also assess the overall similarity of all loops across the four antibodies:

```{r}
# Compare all loops with weights for four antibodies
OverallSimilarity <- ReadAb::AssessOverallLoopSimilarity(
  list(
    "Antibody1" = ab1,
    "Antibody2" = ab2,
    "Antibody3" = ab3,
    "Antibody4" = ab4
  ),
  wH1 = 0.1,
  wH2 = 0.1,
  wH3 = 0.5,
  wL1 = 0.1,
  wL2 = 0.1,
  wL3 = 0.1
)

# Display the similarity plot
ReadAb::DisplaySimilarityPlot(OverallSimilarity, loop = 'all')
```

### Conclusion

By following these steps, you can assess the similarity of specific loops (or all loops) across multiple antibodies. The ReadAb package provides powerful tools for analyzing antibody structures, comparing their binding loops, and visualizing the results with 3D structures and heatmaps.

## Package Refernces
- Gillis, C (2024) ReadAb: Antibody Reading, Visualization, and Loop Comparison Toolkit. Unpublished. https://github.com/claregillis/ReadAb.git

<br>

## Other References

- Chiu, M. L., Goulet, D. R., Teplyakov, A., & Gilliland, G. L. (2019). Antibody structure and function: The basis for engineering therapeutics. *Antibodies* (Basel), 8(4), Article 55. https://doi.org/10.3390/antib8040055

- Chothia, C., Lesk, A. M., Tramontano, A., Levitt, M., Smith-Gill, S. J., Air, G., et al. (1989). Conformations of immunoglobulin hypervariable regions. *Nature*, 342:6252, 877–883. Available from: [https://www.nature.com/articles/342877a0](https://www.nature.com/articles/342877a0)

- Dondelinger, M., Filée, P., Sauvage, E., Quinting, B., Muyldermans, S., Galleni, M., & Vandevenne, M. S. (2018). Understanding the significance and implications of antibody numbering and antigen-binding surface/residue definition. *Frontiers in Immunology*, 9, Article 2278. https://doi.org/10.3389/fimmu.2018.02278

- Dunbar, J., Krawczyk, K., Leem, J., Baker, T., Fuchs, A., Georges, G., Shi, J., & Deane, C. M. (2014). SAbDab: The structural antibody database. *Nucleic Acids Research*, 42(D1), D1140–D1146. https://doi.org/10.1093/nar/gkt1043

- Grant, B. J., Skjaerven, L., & Yao, X. Q. (2021). The Bio3D packages for structural bioinformatics. *Protein Science*, 30(1), 20–30. https://doi.org/10.1002/pro.3923

- Honegger, A., Plückthun, A. (2001). Yet Another Numbering Scheme for Immunoglobulin Variable Domains: An Automatic Modeling and Analysis Tool. *J Mol Biol*, 309(3), 657–670.

- Kabat EA, Te Wu T, Bilofsky H, (U.S.) NI of H. Sequences of Immunoglobulin Chains: Tabulation and Analysis of Amino Acid Sequences of Precursors, V-regions, C-regions, J-Chain and BP-Microglobulins, 1979. Department of Health, Education, and Welfare, Public Health Service, National Institutes of Health (1979). Available online at: [https://books.google.com/books?id=OpW8-ibqyvcC](https://books.google.com/books?id=OpW8-ibqyvcC)

- Lefranc, M. P., Lefranc, G. (2020). Immunoglobulins or Antibodies: IMGT® Bridging Genes, Structures and Functions. *Biomedicines*, 8(9):319. Available from: [https://www.mdpi.com/2227-9059/8/9/319/htm](https://www.mdpi.com/2227-9059/8/9/319/htm)

- Pagès, H., Aboyoun, P., Gentleman, R., & DebRoy, S. (2024). Biostrings: Efficient manipulation of biological strings (R package version 2.72.1). https://doi.org/10.18129/B9.bioc.Biostrings

- Sievert, C. (2020). Interactive Web-Based Data Visualization with R, plotly, and shiny. Chapman and Hall/CRC Florida.

- Wickham, H. (2007). Reshaping data with the reshape package. *Journal of Statistical Software*, 21(12).

- Wickham H, François R, Henry L, Müller K, Vaughan D (2023). *dplyr: A Grammar of Data Manipulation*. R package version 1.1.4. [https://CRAN.R-project.org/package=dplyr](https://CRAN.R-project.org/package=dplyr)

- Wright, E. S. (2016). “Using DECIPHER v2.0 to Analyze Big Biological Sequence Data in R.” *The R Journal*, *8*(1), 352-359.

----

```{r}
sessionInfo()
```

# [END]